pcsx2/R5900.cpp

```diff
@@ -571,44 +686,73 @@ void __fastcall eeloadHook()
 				elftoload = discelf.ToUTF8();
 		}
 
+		// When fast-booting, we insert the game's ELF name into EELOAD so that the game is called instead of the default call of
+		// "rom0:OSDSYS"; any launch arguments supplied by the user will be inserted into EELOAD later by eeloadHook2()
 		if (!elftoload.empty())
 		{
-#if 0
-			// FIXME: you'd think that changing argc and argv would work but no, need to work out why not
-			// This method would support adding command line arguments to homebrew (and is generally less hacky)
-			// It works if you hook on rom0:PS2LOGO loading, but not on the first call with no arguments
-			cpuRegs.GPR.n.a0.SD[0] = 2; // argc = 2
-			// argv[0] = "EELOAD"
-			strcpy((char*)PSM(cpuRegs.GPR.n.a1.UD[0] + 0x40), "EELOAD");
-			memWrite32(cpuRegs.GPR.n.a1.UD[0] + 0, cpuRegs.GPR.n.a1.UD[0] + 0x40);
-			// argv[1] = elftoload
-			strcpy((char*)PSM(cpuRegs.GPR.n.a1.UD[0] + 0x47), elftoload.c_str());
-			memWrite32(cpuRegs.GPR.n.a1.UD[0] + 4, cpuRegs.GPR.n.a1.UD[0] + 0x47);
-			memWrite32(cpuRegs.GPR.n.a1.UD[0] + 8, 0);
-			g_GameLoading = true;
-			return;
-#else
-			// The strings are all 64-bit aligned.  Why? I don't know, but they are
-			for (u32 osdsys_str = EELOAD_START; osdsys_str < EELOAD_START + EELOAD_SIZE; osdsys_str += 8) {
-				if (!strcmp((char*)PSM(osdsys_str), "rom0:OSDSYS")) {
-					for (u32 osdsys_ptr = osdsys_str - 4; osdsys_ptr >= EELOAD_START; osdsys_ptr -= 4) {
-						if (memRead32(osdsys_ptr) == osdsys_str) {
-							strcpy((char*)PSM(cpuRegs.GPR.n.a1.UD[0] + 0x40), elftoload.c_str());
-							memWrite32(osdsys_ptr, cpuRegs.GPR.n.a1.UD[0] + 0x40);
-							g_GameLoading = true;
-							return;
-						}
-					}
+			// Find and save location of default/fallback call "rom0:OSDSYS"; to be used later by eeloadHook2()
+			for (g_osdsys_str = EELOAD_START; g_osdsys_str < EELOAD_START + EELOAD_SIZE; g_osdsys_str += 8) // strings are 64-bit aligned
+			{
+				if (!strcmp((char*)PSM(g_osdsys_str), "rom0:OSDSYS"))
+				{
+					// Overwrite OSDSYS with game's ELF name
+					strcpy((char*)PSM(g_osdsys_str), elftoload.c_str());
+					g_GameLoading = true;
+					return;
 				}
 			}
-#endif
 		}
 	}
 
 	if (!g_GameStarted && (disctype == 2 || disctype == 1) && elfname == discelf)
 		g_GameLoading = true;
 }
 
+// Called from recompilers; __fastcall define is mandatory.
+// Only called if g_SkipBiosHack is true
+void __fastcall eeloadHook2()
+{
+	if (g_Conf->CurrentGameArgs.empty())
+		return;
+
+	if (!g_osdsys_str)
+	{
+		Console.WriteLn("eeloadHook2: Called before \"rom0:OSDSYS\" was found by eeloadHook()!");
+		return;
+	}
+
+#if DEBUG_LAUNCHARG
+	const char *argString = g_Conf->CurrentGameArgs.c_str();
+	Console.WriteLn("eeloadHook2: Attempting to supply launch argument(s) '%s' to ELF '%s'.", argString, (char *)PSM(g_osdsys_str));
```

Comment:
At least one of these DEBUG_LAUNCHARG only messages should always show, so users see that it's being attempted.
If it's not this one, please choose the one that's most likely to cause issues.

Thanks! :)

Reply:
Thanks for the encouragement. You have some good points and I will make commits soon.  I'll enable one message in each hook function -- one for the full boot arg injection and one for the fast boot injection (which would be the message above).

Reply:
Great!